ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp

```diff
@@ -625,6 +640,13 @@ static int HandleFileOpOperation(
         if (!TryGetFileIsFlaggedAsInRoot(currentVnode, context, &fileFlaggedInRoot))
         {
             KextLog_Info("Failed to read attributes when handling FileOp operation. Path is %s", path);
+
+            const char* vnode_name = vnode_getname(currentVnode);
+            KextLog_Error("KAUTH_FILEOP_CLOSE: checking file flags failed. Path = '%s' Vnode name: %s, type %d, being recycled: %s",
+                path, vnode_name ?: "[NULL]", vnode_vtype(currentVnode), vnode_isrecycled(currentVnode) ? "yes" : "no");
```

Comment:
Should we also check if `path` is null?  I think you might have mentioned seeing a null path when the vnode had been recycled.

Reply:
The `vsprintf` implementation in the kernel already is hardened against `NULL` arguments for `%s`, so I don't see a need to redo this explicitly.

Reply:
> The vsprintf implementation in the kernel already is hardened against NULL arguments for %s, so I don't see a need to redo this explicitly.

How does that implementation behave when it encounters a NULL?  What prompted this question was the NULL included for `vnode_name`:

```vnode_name ?: "[NULL]"```

Reply:
Ah yes, that's a little inconsistent; I just checked, and in practice, `path` [(arg1) will *never* be `NULL` for `KAUTH_FILEOP_CLOSE`](https://github.com/apple/darwin-xnu/blob/a449c6a3b8014d9406c2ddbdc81795da24aa7443/bsd/kern/kern_authorization.c#L561).

Incidentally, I believe the issue I noticed was that `vn_getpath()` was failing/returning an empty path when called during `KextLog_File*()` in an error message.

Reply:
To answer your question: the built-in hardening [treats `NULL` strings as the empty string](https://github.com/apple/darwin-xnu/blob/a449c6a3b8014d9406c2ddbdc81795da24aa7443/osfmk/kern/printf.c#L443) which isn't exactly pretty and (to me) doesn't stand out when scanning error messages.

As it looks like path will never be NULL anyway, adding mitigation doesn't really make any sense, and likewise I think it might be useful to specifically call out the case of a NULL vnode name, so I'm inclined to stick with the current version of the code.